core/src/main/java/com/datastax/oss/driver/api/core/tracker/RequestTracker.java

```diff
@@ -148,16 +204,17 @@ default void onNodeSuccess(
    * @param executionProfile the execution profile of this request.
    * @param node the node that returned the successful response.
    * @param requestLogPrefix the dedicated log prefix for this request
+   * @param executionInfo the execution info containing the results of this request
    */
   default void onNodeSuccess(
       @NonNull Request request,
       long latencyNanos,
       @NonNull DriverExecutionProfile executionProfile,
       @NonNull Node node,
-      @NonNull String requestLogPrefix) {
-    // If client doesn't override onNodeSuccess with requestLogPrefix delegate call to the old
-    // method
-    onNodeSuccess(request, latencyNanos, executionProfile, node);
+      @NonNull String requestLogPrefix,
+      @NonNull ExecutionInfo executionInfo) {
+    // delegate call to the old method
+    onNodeSuccess(request, latencyNanos, executionProfile, node, requestLogPrefix);
```

Comment:
IIRC, from what I could tell, in most error cases no ExecutionInfo is present (but I'm still new to the driver code).  I've started going through it again and I think CqlRequestHandler can be refactored to pass it in for some cases.  I'll add the new arg without the NonNull annotation.

Reply:
CqlRequestHandler is a bit tangled - there's a NodeResponseCallback inner class with a trackNodeError method and on the containing class, there's a setFinalError method where the ExecutionInfo is created.
Ideally, setFinalError could call RequestTracker.setNodeError with the ExecutionInfo, but that's already being done in NodeResponseCallback.trackNodeError so I'm thinking it's appropriate to move the trackNodeError method to the containing class.

Reply:
I agree that CqlRequestHandler does get a little tangled; async code certainly will do that if you don't keep an eye on it :).

My comment wasn't connected to the current implementation of CqlRequestHandler, though.  Individual request handlers are created by the various request processors.  These processors vary widely in implementation, and that's just for the ones we've written; they're defined by an interface so users are always free to write their own!  With that in mind I'd say it's important to think about RequestTracker as a general interface which may (or may not) be used by any given handler for any given processor.

With that in mind, I'd argue it's important for RequestTracker to present a coherent interface that's consistent across all the on*() handler methods unless there's a _very_ good reason to diverge.  There doesn't appear to be any such reason in this case, thus my recommendation to update onNodeError() to keep things consistent.

Reply:
I totally agree on keeping the interface consistent.

I've been going through the handlers and for the most part, calls to onNodeError with a null ExecutionInfo are feasible, so that's a good first step.  Should a larger refactoring effort to route the ExecutionInfo to those methods be postponed?  I suppose a good argument against the refactoring work is that the node-level events in the interface can be considered "in-progress" requests and the final ExecutionInfo instance will be delivered in onSuccess.

The only place I see where we can't honor the NonNull contract for onSuccess without significant changes is ContinuousRequestHandlerBase.

Reply:
I figured out a solution in `ContinuousRequestHandlerBase`.
https://github.com/datastax/java-driver/pull/1640/files#r1247216557

Reply:
Can we do this in `CqlRequestHandler.NodeResponseCallback.trackNodeError`?
```java
    private void trackNodeError(Node node, Throwable error, long nodeResponseTimeNanos) {
      if (requestTracker instanceof NoopRequestTracker) {
        return;
      }
      if (nodeResponseTimeNanos == NANOTIME_NOT_MEASURED_YET) {
        nodeResponseTimeNanos = System.nanoTime();
      }
      long latencyNanos = nodeResponseTimeNanos - this.nodeStartTimeNanos;
      ExecutionInfo executionInfo =
              new DefaultExecutionInfo(
                      statement,
                      node,
                      startedSpeculativeExecutionsCount.get(),
                      execution,
                      errors,
                      null,
                      null,
                      true,
                      session,
                      context,
                      executionProfile);
      requestTracker.onNodeError(
          statement, error, latencyNanos, executionProfile, node, logPrefix, executionInfo);
    }
```

Reply:
I also explored how to pass in ExecutionInfo from ContinuousRequestHandlerBase, see [here](https://github.com/SiyaoIsHiding/java-driver/compare/4.x...SiyaoIsHiding:java-driver:request-tracker-execution-info?expand=1). How do you think of this approach?

Reply:
Changes to `ContinuousRequestHandlerBase` merged.  Thank you.